src/nvapi_disp.cpp

```diff
@@ -96,15 +89,85 @@ extern "C" {
+                    pHDRColorDataV1->mastering_display_data.displayPrimary_x0 = output->GetColorData().RedPrimaryX;
+                    pHDRColorDataV1->mastering_display_data.displayPrimary_y0 = output->GetColorData().RedPrimaryY;
+                    pHDRColorDataV1->mastering_display_data.displayPrimary_x1 = output->GetColorData().GreenPrimaryX;
+                    pHDRColorDataV1->mastering_display_data.displayPrimary_y1 = output->GetColorData().GreenPrimaryY;
+                    pHDRColorDataV1->mastering_display_data.displayPrimary_x2 = output->GetColorData().BluePrimaryX;
+                    pHDRColorDataV1->mastering_display_data.displayPrimary_y2 = output->GetColorData().BluePrimaryY;
+                    pHDRColorDataV1->mastering_display_data.displayWhitePoint_x = output->GetColorData().WhitePointX;
+                    pHDRColorDataV1->mastering_display_data.displayWhitePoint_y = output->GetColorData().WhitePointY;
+                    pHDRColorDataV1->mastering_display_data.max_display_mastering_luminance = output->GetColorData().MaxLuminance;
+                    pHDRColorDataV1->mastering_display_data.min_display_mastering_luminance = output->GetColorData().MinLuminance;
+                    pHDRColorDataV1->mastering_display_data.max_content_light_level = output->GetColorData().MaxLuminance;
+                    pHDRColorDataV1->mastering_display_data.max_frame_average_light_level = output->GetColorData().MaxFullFrameLuminance;
+                }
+            } else {
+                DXGI_COLOR_SPACE_TYPE colorspace = pHDRColorDataV1->hdrMode != NV_HDR_MODE_OFF ? DXGI_COLOR_SPACE_RGB_FULL_G2084_NONE_P2020 : DXGI_COLOR_SPACE_RGB_FULL_G22_NONE_P709;
+                DXGI_HDR_METADATA_HDR10 metadata = {
+                    .RedPrimary = { pHDRColorDataV1->mastering_display_data.displayPrimary_x0, pHDRColorDataV1->mastering_display_data.displayPrimary_y0 },
+                    .GreenPrimary = { pHDRColorDataV1->mastering_display_data.displayPrimary_x1, pHDRColorDataV1->mastering_display_data.displayPrimary_y1 },
+                    .BluePrimary = { pHDRColorDataV1->mastering_display_data.displayPrimary_x2, pHDRColorDataV1->mastering_display_data.displayPrimary_y2 },
+                    .WhitePoint = { pHDRColorDataV1->mastering_display_data.displayWhitePoint_x, pHDRColorDataV1->mastering_display_data.displayWhitePoint_y },
+                    .MaxMasteringLuminance = pHDRColorDataV1->mastering_display_data.max_display_mastering_luminance,
+                    .MinMasteringLuminance = pHDRColorDataV1->mastering_display_data.min_display_mastering_luminance,
+                    .MaxContentLightLevel = pHDRColorDataV1->mastering_display_data.max_content_light_level,
+                    .MaxFrameAverageLightLevel = pHDRColorDataV1->mastering_display_data.max_frame_average_light_level,
+                };
+                if (FAILED(interop->SetGlobalHDRState(colorspace, &metadata)))
```

Comment:
How global should this state be, exactly? Since this NVAPI function operates only on one display, shouldn't this interface extend something like `IDXGIOutput` instead? Or does setting HDR state on one display makes it retrievable from all displays even on Windows with native NVAPI?

Reply:
Swapchains are not necessarily linked to outputs.

We treat it as a global override for all of the app's swapchains to be treated as HDR, etc.

This is obviously different to the behaviour on Windows where it forces a single display into a weird forced HDR mode, but it ends up being better overall.

Reply:
That makes sense. Let's just hope that we won't confuse any app by implementing it this way.